### PR TITLE
Add THD + GQA supports

### PR DESCRIPTION
**Pull request by zlsh80826**
# Description

Support GQA/MHA with THD format

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

- Support GQA with THD format
- Add GQA + THD JAX unit tests

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by mgoldfarb-nvidia — APPROVED** [review]
LGTM

**Comment by zlsh80826** [comment]
Hi @cyanguwa, the CI failures seem not related to this PR, could you help take a look if this PR is ready to be merged? Thanks!

File: tests/jax/test_fused_attn.py [hunk 1]
```diff
@@ -305,11 +305,14 @@ def _check_configs(self):
...
             pytest.skip("THD format requires padding masks.")
 
-        if self.qkv_layout == QKVLayout.BS3HD or get_qkv_format(self.qkv_layout) == QKVFormat.THD:
-            if self.num_heads_q != self.num_heads_kv:
-                pytest.skip("QKVPACKED layout requires num_heads_q and num_heads_kv to be equal.")
+        qkv_format = get_qkv_format(self.qkv_layout)
+        if self.qkv_layout == QKVLayout.BS3HD or qkv_format == QKVFormat.THD:
             if self.max_seqlen_q != self.max_seqlen_kv:
-                pytest.skip("QKVPACKED layout requires max_seqlen_q and max_seqlen_kv to be equal.")
+                pytest.skip(f"{self.qkv_layout} requires max_seqlen_q == max_seqlen_kv")
```
Comment (cyanguwa) [inline tests/jax/test_fused_attn.py:311]: Why do we skip cross attention here for THD? It should work for THD_T2HD and THD_TH2D cases as well right?

Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:311]: Nice catch, PMed you.

Comment (cyanguwa) [reply tests/jax/test_fused_attn.py:311]: Thanks for the details. We'll merge the PR and wait for the cross-attention related bug to be fixed then. Thanks.

**Pull request merged**